find_tags: Use a hashtab_T instead of garray_T to collect tag results

jamessan opened the pull request:
As discussed in #1044, much of the time currently spent in finding tags
is looking for duplicate tags.  Since the tags are stored in a garray_T,
this has to perform a linear search of all existing tags every time a
new tag is added.

Using a hashtab_T (with a modified hash function to allow embedded NUL
bytes), the deduplication costs are drastically cheaper.

Comparing the times to find all tags containing the letter a in a tags
file for all of /usr/include:

``` sh
  $ time ./src/vim.hash -u NONE --cmd 'echo len(taglist("a"))' --cmd q
  115305
  ./src/vim -u NONE --cmd 'echo len(taglist("a"))' --cmd q  0.69s user 0.04s system 98% cpu 0.741 total
  $ time ./src/vim -u NONE --cmd 'echo len(taglist("a"))' --cmd q
  115305
  ./src/vim -u NONE --cmd 'echo len(taglist("a"))' --cmd q  80.69s user 0.08s system 99% cpu 1:20.90 total
```

jamessan committed:
find_tags: Use a hashtab_T instead of garray_T to collect tag results

As discussed in #1044, much of the time currently spent in finding tags
is looking for duplicate tags.  Since the tags are stored in a garray_T,
this has to perform a linear search of all existing tags every time a
new tag is added.

Using a hashtab_T (with a modified hash function to allow embedded NUL
bytes), the deduplication costs are drastically cheaper.

Comparing the times to find all tags containing the letter a in a tags
file for all of /usr/include:

  $ time ./src/vim.hash -u NONE --cmd 'echo len(taglist("a"))' --cmd q
  115305
  ./src/vim -u NONE --cmd 'echo len(taglist("a"))' --cmd q  0.69s user 0.04s system 98% cpu 0.741 total
  $ time ./src/vim -u NONE --cmd 'echo len(taglist("a"))' --cmd q
  115305
  ./src/vim -u NONE --cmd 'echo len(taglist("a"))' --cmd q  80.69s user 0.08s system 99% cpu 1:20.90 total

khalidchawtany commented:
That is a hell lot faster 😍

jamessan commented:
I need to do some additional work to properly support non-NUL terminated strings as the hash key, but I think this demonstrates that there's significant benefit in switching away from garray_T to accumulate the tags.

nuko8 commented on src/tag.c:
```diff
@@ -2247,179 +2267,159 @@ find_tags(
-		    int len;
-		    int heuristic;
+		int len;
+		int heuristic;
```
Looks the declarations of `len` and `heuristic` need to be moved somewhere (just after the left curly brace at line 2217, maybe?) to comply with ANSI/ISO C90.  Sorry about my being `-pedantic` :)

FWIW, I tried this patch against 7.4.2342 on OS X 10.11.6 (El Capitan) with the same configuration as [the failed CI test](https://travis-ci.org/vim/vim/jobs/158148466); the test fully succeeded for me.

jamessan replied:
Thanks for the feedback!  I initially had those in a nested block (just removed the if condition).  Now that I've removed that, I do need to move the declarations.

As far as standard compatibility, it might be worth adding a CI build that uses `-std=gnu89 -pedantic` to more easily find these issues.  I notice that's not the only issue in the current source.

jamessan committed:
hashtab: Add support for using fixed-length buffers as keys

jamessan commented:
`hashtab_T` can now handle fixed-length buffers of data as keys.  There are new `hash_buf_*` functions which have an extra keylen argument to describe the size of the buffer.  The existing `hash_*` functions have been changed to be wrappers around the `hash_buf_*` functions, calling `STRLEN(key)` as needed.

jamessan committed:
Separate key length from number of bytes used to compare tags

This restores the original behavior of tag comparisons, fixing the
Test_help_complete test failures.

jamessan commented:
@brammool This is done from my end, pending any further comments.

jamessan commented:
@chrisbra Something seems to be up with the mirroring to vim-dev.  None of my comments in this post nor some of the recent PRs/issues have shown up on the list.

chrisbra commented:
Am 2016-09-14 04:03, schrieb James McCoy:

> @chrisbra [1] Something seems to be up with the mirroring to vim-dev.
> None of my comments in this post nor some of the recent PRs/issues
> have shown up on the list.

Yeah, I noticed that too. There were several notifications from 
yesterday,
that did not come back to vim-dev.

I think that mirror email has been accidentally removed from posting to 
the vim-dev
list because of the recent spam that was visible there. I'll activate it 
later
today.

Best,
Christian

jamessan commented:
Updated against 8.0.0046.

jamessan commented:
Updated against 8.0.0082.

jamessan commented:
Updated against 8.0.0176.
